ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarRecoveringCustodyImpl.java

```diff
@@ -0,0 +1,290 @@
+import tech.pegasys.teku.kzg.KZG;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.SpecMilestone;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.fulu.DataColumnSidecar;
+import tech.pegasys.teku.spec.datastructures.blocks.BeaconBlock;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.blocks.SlotAndBlockRoot;
+import tech.pegasys.teku.spec.datastructures.networking.libp2p.rpc.DataColumnIdentifier;
+import tech.pegasys.teku.spec.datastructures.util.DataColumnSlotAndIdentifier;
+import tech.pegasys.teku.spec.logic.versions.fulu.helpers.MiscHelpersFulu;
+import tech.pegasys.teku.statetransition.blobs.BlobSidecarManager;
+
+public class DataColumnSidecarRecoveringCustodyImpl implements DataColumnSidecarRecoveringCustody {
+  private static final Logger LOG = LogManager.getLogger("das-nyota");
+
+  private final DataColumnSidecarByRootCustody delegate;
+  private final AsyncRunner asyncRunner;
+  private final MiscHelpersFulu miscHelpers;
+  private final KZG kzg;
+  private final Spec spec;
+
+  private final long columnCount;
+  private final int recoverColumnCount;
+  private final boolean isSuperNode;
+
+  final Function<UInt64, Duration> slotToRecoveryDelay;
+  private final Map<SlotAndBlockRoot, RecoveryTask> recoveryTasks;
+
+  private final Subscribers<DataColumnSidecarManager.ValidDataColumnSidecarsListener>
+      validDataColumnSidecarsSubscribers = Subscribers.create(true);
+
+  public DataColumnSidecarRecoveringCustodyImpl(
+      final DataColumnSidecarByRootCustody delegate,
+      final AsyncRunner asyncRunner,
+      final Spec spec,
+      final MiscHelpersFulu miscHelpers,
+      final KZG kzg,
+      final boolean isSuperNode,
+      final int columnCount,
+      final Function<UInt64, Duration> slotToRecoveryDelay) {
+    this.delegate = delegate;
+    this.asyncRunner = asyncRunner;
+    this.miscHelpers = miscHelpers;
+    this.kzg = kzg;
+    this.spec = spec;
+    this.recoveryTasks =
+        LimitedMap.createSynchronizedNatural(spec.getGenesisSpec().getSlotsPerEpoch());
+    this.isSuperNode = isSuperNode;
+    this.slotToRecoveryDelay = slotToRecoveryDelay;
+    this.columnCount = columnCount;
+    this.recoverColumnCount = columnCount / 2;
+  }
+
+  @Override
+  public void onSlot(final UInt64 slot) {
+    if (!isActiveSuperNode(slot)) {
+      return;
+    }
+    asyncRunner
+        .runAfterDelay(
+            () -> {
+              LOG.debug("Check if recovery needed for slot: {}", slot);
+
+              final List<SlotAndBlockRoot> ids =
+                  recoveryTasks.keySet().stream()
+                      .filter(key -> key.getSlot().equals(slot))
+                      .toList();
+              ids.stream()
+                  .map(recoveryTasks::get)
+                  .forEach(
+                      task -> {
+                        task.timeouted().set(true);
```

Comment:
nit: as an option instead of `timedOut` field we may just have class wide `timedOutSlot` field, increase it here and then consider all tasks with slots below as time out

Reply:
I want recovery start at certain seconds in a slot

Reply:
It would, if you update `timedOutSlot` and then call `maybeStartRecovery()`. 
But anyway the current version also works, just a bit of excessive data

Reply:
So I could have 64+ data columns already but don't want to fire recovery if it's just 1 second into slot

Reply:
I could do only this as an improvement https://github.com/Nashatyrev/teku/pull/191/commits/d2447057d147e810740262f3e82936189e5f056b